pinot-core/src/main/java/org/apache/pinot/core/accounting/PerQueryCPUMemAccountantFactory.java

```diff
@@ -724,14 +745,14 @@ void killAllQueries() {
        */
       private void killMostExpensiveQuery() {
         if (!_aggregatedUsagePerActiveQuery.isEmpty() && _numQueriesKilledConsecutively >= _gcTriggerCount) {
-          System.gc();
           _numQueriesKilledConsecutively = 0;
+          System.gc();
           try {
-            Thread.sleep(_normalSleepTime);
+            Thread.sleep(_gcWaitTime);
           } catch (InterruptedException ignored) {
           }
           _usedBytes = MEMORY_MX_BEAN.getHeapMemoryUsage().getUsed();
-          if (_usedBytes < _criticalLevel) {
+          if (_usedBytes < _criticalLevelAfterGC) {
```

Comment:
Would this end up killing more queries than usual when just killing one query would have been sufficient?   

Let's say 
criticalThreshold=95%
usedMemory=96%
CriticalLevelAFterGC=80%

After performing GC in L749, if the usedMemory=85%, we would end up killing the most expensive query even though the criticalThreshold is 95%. Is that expected?

Reply:
Yes that is expected to prevent heap size oscillation and repetitive gc triggering. So in production this code path is almost never triggered as `_numQueriesKilledConsecutively` is mostly at a low level and `_numQueriesKilledConsecutively >= _gcTriggerCount` is false. We now want to also use this path to "give gc a second chance" if it hasn't kick in by setting _gcTriggerCount to 0. For these use cases, even HeapMemoryCritical event is very seldom, so when it happens we want to make sure it first to a gc. What we observed there is once the gc happens it would collect a ton of garbage. So if it fails to do so (`_usedBytes > _criticalLevelAfterGC`) we should still kill the query. Nevertheless, this _criticalLevelAfterGC itself is configurable so we can make `_criticalLevelAfterGC= _criticalLevel` if we want the detection to be even less sensitive.

Reply:
Understood. Does that mean the default for delta config variable should be 0?

Reply:
We can set it to default to 5. 0 is a special case to enable the always gc before kill behaviour. I have added related descriptions in the code comment. Will add them to git doc as well.